### PR TITLE
Enable ruff flake8 use pathlib (pth)

### LINKED ISSUE #13295: Applying more Ruff groups to this repository
The current selection of Ruff rules comes as a direct translation from when we were using isort, pyupgrade, pycln, Flake8, and some plugins.
It made perfect sense at the time to restrict ourselves to only the most necessary rules, as additional linters and tooling to be installed and run could mean an increase in CI runtime, security risks, upgrade maintenance, and complications between incompatible versions of libraries.

But with Ruff, the context is completely different. It is so fast than any additional rule group is negligible on the runtime and it's a single dependency. Which means that we could enable more rule groups w/o much concerns to help keep our growing non-stubs code up to the best standards. And explore more lints that could be beneficial to stubs.

As a side effect, we get to test various Ruff rules on a very 3rd-party type-stubs-driven repository, helping catch edge-cases and rules that don't belong in that context. This, in my opinion, would further bolster the existing collaboration between Ruff and typeshed, as well as indirectly help the Python typing community that also uses Ruff.

---

I'm already ready with an experiment PR to see the width and scale of changes this would entail with nearly all rules enabled.
https://github.com/Avasam/typeshed/pull/39 (not something to be reviewed, just a launching point for me, like a possible "todo list")
Of course I would go incrementally: starting with those that already pass as-is, then those covered by autofixes, individually splitting off the more massive ones, finishing with the manual interventions.

---

<details>
<summary>Checklist</summary>

- `Flake8`:
  - `Pyflakes (F)`: https://github.com/python/typeshed/pull/11496
  - `pycodestyle Error (E)`: https://github.com/python/typeshed/pull/11496
  - `pycodestyle Warning (W)`: https://github.com/python/typeshed/pull/11496
  - `mccabe (C90)`: ❌We don't actually mind long and complex tests. This is also very arbitrary, let's make that call on individual reviews
- `isort (I)`: https://github.com/python/typeshed/pull/10912
- `pep8-naming (N)`: https://github.com/python/typeshed/pull/13327
- `pydocstyle (D)`: https://github.com/python/typeshed/pull/13326 & 
- `pyupgrade (UP)`: https://github.com/python/typeshed/pull/11499
- `flake8-2020 (YTT)`: https://github.com/python/typeshed/pull/13314
- `flake8-annotations (ANN)`: https://github.com/python/typeshed/pull/13331
  - ⚠Once https://github.com/python/typeshed/issues/9550 is complete, `ANN401` *could* be considered
- `flake8-async (ASYNC)`: https://github.com/python/typeshed/pull/13727
- `flake8-bandit (S)`: https://github.com/Avasam/typeshed/tree/Ruff-S-flake8-bandit
- `flake8-blind-except (BLE)`: https://github.com/python/typeshed/pull/13728
- `flake8-boolean-trap (FBT)`: ☑https://github.com/python/typeshed/pull/13751
- `flake8-bugbear (B)`: https://github.com/python/typeshed/pull/11500 & https://github.com/python/typeshed/pull/13308
- `flake8-builtins (A)`: https://github.com/python/typeshed/pull/13729
- `flake8-commas (COM)`: ❌Already taken care of by, or conflict with, Black
- `flake8-copyright (CPY)`: ❌Not using top-of-file copyright notices
- `flake8-comprehensions (C4)`: ☑https://github.com/python/typeshed/pull/13751
- `flake8-datetimez (DTZ)`: ☑https://github.com/python/typeshed/pull/13751
- `flake8-debugger (T10)`: ☑https://github.com/python/typeshed/pull/13751
- `flake8-django (DJ)`: ❌Not using Django at runtime
- `flake8-errmsg (EM)`: https://github.com/python/typeshed/pull/13730
- `flake8-executable (EXE)`: https://github.com/python/typeshed/pull/13346
- `flake8-future-annotations (FA)`: https://github.com/python/typeshed/pull/10910
- `flake8-implicit-str-concat (ISC)`: ❌Mostly taken care of by the formatter. And we allow both implicit and explicit multiline strings (to my dismay 😆)
- `flake8-import-conventions (ICN)`: https://github.com/python/typeshed/pull/13731
- `flake8-logging (LOG)`: ❌Not using logging at runtime
- `flake8-logging-format (G)`: ❌Not using logging at runtime
- `flake8-no-pep420 (INP)`: ❌Leave the decision at review to use [PEP 420](https://peps.python.org/pep-0420/)
- `flake8-pie (PIE)`: https://github.com/python/typeshed/pull/13747
- `flake8-print (T20)`: ❌We use print as a logging mechanism
- `flake8-pyi (PYI)`: https://github.com/python/typeshed/pull/10458 and ongoing
- `flake8-pytest-style (PT)`: ❌Not using pytest at runtime / in our test suite
- `flake8-quotes (Q)`: ❌This should be fully covered by formatters (whether Black or Ruff)
- `flake8-raise (RSE)`: ☑https://github.com/python/typeshed/pull/13751
- `flake8-return (RET)`: https://github.com/python/typeshed/pull/13323
- `flake8-self (SLF)`: ❌Already configured to ignore in stubs and covered by [pyright's `reportprivateusage`](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportprivateusage).
- `flake8-slots (SLOT)`: ☑https://github.com/python/typeshed/pull/13751
- `flake8-simplify (SIM)`: https://github.com/python/typeshed/pull/13309
- `flake8-tidy-imports (TID)`: ❌Banned APIs need to be configured to be useful. A mix of stubs use relative imports.
- `flake8-type-checking (TC)`: https://github.com/python/typeshed/pull/13333
- `flake8-gettext (INT)`: ❌Not using gettext or translations
- `flake8-unused-arguments (ARG)`: https://github.com/python/typeshed/pull/13334
- `flake8-use-pathlib (PTH)`: https://github.com/python/typeshed/pull/13750 & 
- `flake8-todos (TD)`: https://github.com/python/typeshed/pull/13748
- `flake8-fixme (FIX)`: ❌It doesn't make sense to block the CI on TODOs
  - ⚠There's a feature request to configure rules as warnings in Ruff: https://github.com/astral-sh/ruff/issues/1256
- `eradicate (ERA)`: ❌Way too many false-positives
- `pandas-vet (PD)`: ❌Not using Pandas at runtime
- `pygrep-hooks (PGH)`: https://github.com/python/typeshed/pull/13304
- `Pylint (PL)`
  - `Pylint Convention (PLC)`: https://github.com/python/typeshed/pull/13306
  - `Pylint Error (PLE)`: https://github.com/python/typeshed/pull/13305
  - `Pylint Refactor (PLR)`: https://github.com/python/typeshed/pull/13307
  - `Pylint Warning (PLW)`: https://github.com/python/typeshed/pull/13749
- `tryceratops (TRY)`: https://github.com/python/typeshed/pull/13359
- `flynt (FLY)`: ☑https://github.com/python/typeshed/pull/13751
- `NumPy-specific rules (NPY)`: ❌We may have to reference or support deprecated numpy types at times
- `FastAPI (FAST)`: ❌Not using FastAPI at runtime
- `Airflow (AIR)`: ❌Not using Airflow at runtime
- `Perflint (PERF)`: https://github.com/python/typeshed/pull/13312
- `refurb (FURB)`: https://github.com/python/typeshed/pull/13310
- `pydoclint (DOC)`: ⚠(all in preview, enabling this would do nothing atm)
- `Ruff-specific rules (RUF)`: https://github.com/python/typeshed/pull/11496

☑: We already pass all these (not yet configured) as-is. Can enable together.
</details>

### PR DESCRIPTION
**Pull request by Avasam**
Kept this one for last. Closes #13295
Some rules in this group are forcing some refactoring to use `pathlib`. In some cases, it's a lot cleaner, in others it's debatable.

I unconditionally followed all the rules. Please indicate changes you disagree with/preferred in the old style. I'll revert those and disable the relevant rules.

There's also a few `with read` and `with write` that can be rewritten to be more concise with `Path.read_text` and `Path.write_text`

### CODE REVIEW

File: lib/ts_utils/utils.py [hunk 1]
```diff
@@ -203,7 +203,7 @@ def allowlists(distribution_name: str) -> list[str]:
 
 @functools.cache
 def get_gitignore_spec() -> pathspec.PathSpec:
-    with open(".gitignore", encoding="UTF-8") as f:
+    with Path(".gitignore").open(encoding="UTF-8") as f:
```
Comment (srittau) [inline lib/ts_utils/utils.py:206]: I prefer the simple `open()` in this and similar cases. But it's no deal breaker for me. The rest looks good.

File: scripts/sync_protobuf/tensorflow.py [hunk 2]
```diff
@@ -72,21 +71,21 @@ def post_creation() -> None:
 
     for path in STUBS_FOLDER.rglob("*_pb2.pyi"):
         print(f"Fixing imports in '{path}'")
-        with open(path, encoding="utf-8") as file:
+        with path.open(encoding="utf-8") as file:
             filedata = file.read()
```
Comment (brianschubert) [inline scripts/sync_protobuf/tensorflow.py:75]: In cases like this I'm neutral about `path.open(...)`, but I think `path.read_text()` would be an improvement

File: lib/ts_utils/utils.py [hunk 3]
```diff
@@ -203,7 +203,7 @@ def allowlists(distribution_name: str) -> list[str]:
...
         return pathspec.PathSpec.from_lines("gitwildmatch", f.readlines())
```
Comment (Avasam) [inline lib/ts_utils/utils.py:207]: Unrelated to your above comment or this PR directly. But I just read pathspec's doc and found this:

> There is a specialized class, ``pathspec.GitIgnoreSpec``, which more closely
implements the behavior of **gitignore**. This uses ``GitWildMatchPattern``
pattern by default and handles some edge cases differently from the generic
``PathSpec`` class. ``GitIgnoreSpec`` can be used without specifying the pattern
factory::
> 
> 	>>> spec = pathspec.GitIgnoreSpec.from_lines(spec_text.splitlines())

```suggestion
        return pathspec.GitIgnoreSpec.from_lines(f.readlines())
```

Comment (brianschubert) [reply lib/ts_utils/utils.py:207]: Could also omit the call to `.readlines()`. `GitIgnoreSpec.from_lines` takes an iterable of lines, so passing the file i/o object directly would work. Not sure if there's any reason to prefer one way or another, though

Comment (Avasam) [reply lib/ts_utils/utils.py:207]: You're completely right! I would assume `.readlines()` to just be a waste here since you have to both read the file, then iterate over the list again, instead of streaming it once.

`PathSpec.from_lines` simply does `patterns = [pattern_factory(line) for line in lines if line]`

Comment (Avasam) [reply lib/ts_utils/utils.py:207]: I opened https://github.com/python/typeshed/pull/13797

File: lib/ts_utils/requirements.py [hunk 4]
```diff
@@ -13,14 +12,14 @@
...
-        distributions = os.listdir(STUBS_PATH)
+        distributions = [distribution.name for distribution in STUBS_PATH.iterdir()]
 
     return set(itertools.chain.from_iterable([read_dependencies(distribution).external_pkgs for distribution in distributions]))
 
 
 def get_stubtest_system_requirements(distributions: Iterable[str] = (), platform: str = sys.platform) -> list[str]:
     if not distributions:
-        distributions = os.listdir(STUBS_PATH)
+        distributions = [distribution.name for distribution in STUBS_PATH.iterdir()]
```
Comment (Avasam) [inline lib/ts_utils/requirements.py:22]: This seems like a straight up downgrade.
[os-listdir (PTH208)](https://docs.astral.sh/ruff/rules/os-listdir/#os-listdir-pth208)

Only one other change of `os.listdir` --> `Path.iterdir` seems like an actual improvement. (in `tests/mypy_test.py`)

File: tests/mypy_test.py [hunk 5]
```diff
@@ -472,7 +472,7 @@ def test_third_party_stubs(args: TestConfig, tempdir: Path) -> TestSummary:
     gitignore_spec = get_gitignore_spec()
     distributions_to_check: dict[str, PackageDependencies] = {}
 
-    for distribution in sorted(os.listdir("stubs")):
+    for distribution in sorted([distribution.name for distribution in STUBS_PATH.iterdir()]):
```
Comment (Avasam) [inline tests/mypy_test.py:475]: This seems like a straight up downgrade.
[os-listdir (PTH208)](https://docs.astral.sh/ruff/rules/os-listdir/#os-listdir-pth208)

Only one other change of `os.listdir` --> `Path.iterdir` seems like an actual improvement. (in `add_third_party_files` above)

**Comment by srittau** [comment]
I've gone ahead and merged #13943 for now (extracting `_get_relative()` can be postponed to a subsequent PR) to unblock this PR. Which now has merge conflicts, of course.

File: tests/pytype_test.py [hunk 6]
```diff
@@ -191,37 +188,35 @@ def get_missing_modules(files_to_test: Sequence[str]) -> Iterable[str]:
...
-            if not fi.startswith("stubs/"):
+    missing_modules = {
+        associated_package
+        for distribution in stub_distributions
+        for external_req in read_dependencies(distribution).external_pkgs
+        for associated_package in _get_pkgs_associated_with_requirement(external_req.name)
+    }
+
+    with EXCLUDE_LIST.open() as f:
+        for excluded_file in f:
```
Comment (srittau) [inline tests/pytype_test.py:208]: Nit: I would call `excluded_file` `line` instead. This makes it clear that – at this point – this is just some line that was read from the file and is not necessarily an excluded file (which I would expect to be a `Path`).

File: tests/pytype_test.py [hunk 7]
```diff
@@ -191,37 +188,35 @@ def get_missing_modules(files_to_test: Sequence[str]) -> Iterable[str]:
...
+            if not excluded_file.startswith("stubs/"):
                 # Skips comments, empty lines, and stdlib files, which are in
                 # the exclude list because pytype has its own version.
                 continue
-            unused_stubs_prefix, unused_pkg, mod_path = fi.split("/", 2)  # pyright: ignore[reportUnusedVariable]
-            missing_modules.add(os.path.splitext(mod_path)[0])
+            parts = Path(excluded_file.strip()).with_suffix("").parts
+            missing_modules.add("/".join(parts[2:]))
```
Comment (srittau) [inline tests/pytype_test.py:214]: Considering that we are parsing a file here, which is using a particular format (i.e. POSIX-like) for the paths, and we're just doing basic string manipulation, we shouldn't use `pathlib` to process the lines. I think keeping the existing code here is a better option.

**Review by srittau — CHANGES_REQUESTED** [review]
Thanks, just two comments below.

File: tests/pytype_test.py [hunk 8]
```diff
@@ -191,37 +188,35 @@ def get_missing_modules(files_to_test: Sequence[str]) -> Iterable[str]:
...
+    with EXCLUDE_LIST.open() as f:
+        for line in f:
+            if not line.startswith("stubs/"):
                 # Skips comments, empty lines, and stdlib files, which are in
                 # the exclude list because pytype has its own version.
                 continue
-            unused_stubs_prefix, unused_pkg, mod_path = fi.split("/", 2)  # pyright: ignore[reportUnusedVariable]
-            missing_modules.add(os.path.splitext(mod_path)[0])
+            _ts_subdir, _distribution, module_path = line.split("/", 2)
+            missing_modules.add(module_path.rsplit(".", 1)[0])
```
Comment (Avasam) [inline tests/pytype_test.py:214]: This could also be:
```suggestion
            missing_modules.add(module_path.removesuffix(".pyi"))
```

Comment (srittau) [reply tests/pytype_test.py:214]: I agree, to me as well.

**Review by srittau — APPROVED** [review]

File: tests/pytype_test.py [hunk 9]
```diff
@@ -191,37 +188,35 @@ def get_missing_modules(files_to_test: Sequence[str]) -> Iterable[str]:
...
     return missing_modules
```
Comment (Avasam) [inline tests/pytype_test.py:215]: FWIW, this *could* also be:
```py
    return {
        associated_package
        for distribution in stub_distributions
        for external_req in read_dependencies(distribution).external_pkgs
        for associated_package in _get_pkgs_associated_with_requirement(external_req.name)
    } | {
        # Exclude the distribution and file extension from path
        line.split("/", 2)[2].removesuffix(".pyi")
        for line in EXCLUDE_LIST.read_text().splitlines()
        # Skips comments, empty lines, and stdlib files, which are in
        # the exclude list because pytype has its own version.
        if line.startswith("stubs/")
    }
```

Comment (srittau) [reply tests/pytype_test.py]: I wouldn't want to do fairly extensive changes like this in a PR that's mostly about introducing `Path`. (Also, the code above is a bit too clever for my taste.)

Comment (Avasam) [reply tests/pytype_test.py]: Yeah agreed. I don't wanna push this code-change.

Speaking of changes not *directly* related to introducing Path (or rather, enforcing it). There's still a few changes in the pytype script that were more a result of looking at it more closely

File: tests/pytype_test.py (hunk 8 above)
Comment (Avasam) [reply tests/pytype_test.py]: Oh, haha, I know why this breaks. The suffix is `.pyi\n` 😅

**Review by srittau — APPROVED** [review]
